native/components/message-markdown.react.js

```diff
@@ -0,0 +1,51 @@
+// @flow
+
+import type { StyleSheetOf } from '../themes/colors';
+
+import * as React from 'react';
+import { View } from 'react-native';
+import PropTypes from 'prop-types';
+import * as SimpleMarkdown from 'simple-markdown';
+
+import { style, type Style } from '../themes/markdown-styles';
+import rules from './message-markdown-rules.react';
+
+type Props = {|
+  onLink: (url: string) => Promise<any>,
+  darkStyle?: boolean,
+  children: string,
+|};
+
+class Markdown extends React.PureComponent<Props> {
+  static propTypes = {
+    onLink: PropTypes.func.isRequired,
+    darkStyle: PropTypes.bool,
+    children: PropTypes.string.isRequired,
+  };
+
+  static defaultProps = {
+    darkStyle: false,
+  };
+
+  parse: SimpleMarkdown.Parser;
+  renderer: SimpleMarkdown.ReactOutput;
+  style: StyleSheetOf<Style>;
+
+  render() {
+    const theme = this.props.darkStyle ? 'dark' : 'light';
+    this.style = style(theme);
```

Comment:
`this.style` is an example of derived data that we don't want to recalculate on every render. We talked over video chat previously about the main rules for when we need to memoize derived data:
1. If the data can be identical without being strictly equal (objects can be like this, but primitives like strings cannot), and
2. If the data is expensive to recalculate. In our case we have to redefine all of this stuff - `parser`, `renderer`, etc. - which is quite expensive

When both of these rules apply, we know we need to figure out a way to memoize the derived data. The main ways to do this memoization are:

1. `getDerivedStateFromProps` - this React lifecycle function lets us recalculate derived data before `render`. With this solution, we would keep all the derived data in `this.state`. We would also cache the old value of `this.props.darkStyle` in `this.state`, and then in `getDerivedStateFromProps` compare `darkStyle` to see if it changes. If it changes, we would recalculate the derived data and return an object to update the state
2. `createSelector` - `reselect` is designed for derived data from Redux. In our case our input isn't coming from Redux, but we could just create a selector that takes a different object as input, and construct a new version of that object on every `render`. It looks like that object would just have one property: `darkStyle`. `reselect` would make sure that it only recalculates the derived data if `darkStyle` changes (this is what "memoization" means)
3. React Hooks - `React.useMemo` can handle this for us. We could define `parse`, `renderer`, and `style` each using `React.useMemo`, and pass it `darkStyle` in the "deps array" to make sure `React.useMemo` only recalculates these three when `darkStyle` changes.

React Hooks are my preferred solution here. `getDerivedStateFromProps` is generally not recommended by the React team unless you're implementing a "transitioner" within a class component, and `createSelector` is better for cases where we data that comes from Redux, and we want an easy way to reuse it that works with both class components and hooks.

Can you try reimplementing this component as a hook, and making sure the derived data is memoized?